flocker/node/agents/gce.py

```diff
@@ -515,32 +561,32 @@ def _get_attached_to(self, blockdevice_id):
 
     def detach_volume(self, blockdevice_id):
         attached_to = self._get_attached_to(blockdevice_id)
-        # TODO(mewert): Test this race (something else detaches right at this
-        # point). Might involve putting all GCE interactions behind a zope
-        # interface and then using a proxy implementation to inject code.
         self._do_blocking_operation(
-            self._compute.instances().detachDisk, instance=attached_to,
-            deviceName=blockdevice_id)
+            self._compute.instances().detachDisk,
+            instance=attached_to,
+            deviceName=blockdevice_id,
+            timeout=VOLUME_DETATCH_TIMEOUT)
         return None
 
     def get_device_path(self, blockdevice_id):
-        # TODO(mewert): Verify that we need this extra API call.
         self._get_attached_to(blockdevice_id)
-
-        # TODO(mewert): Verify we can get away returning a symlink here, or
-        # just walk the symlink.
         return FilePath(u"/dev/disk/by-id/google-" + blockdevice_id)
 
     def destroy_volume(self, blockdevice_id):
         try:
-            # TODO(mewert) verify timeouts and error conditions.
             self._do_blocking_operation(
                 self._compute.disks().delete,
-                disk=blockdevice_id
+                disk=blockdevice_id,
+                timeout=VOLUME_DELETE_TIMEOUT,
             )
         except HttpError as e:
             if e.resp.status == 404:
                 raise UnknownVolume(blockdevice_id)
+            elif e.resp.status == 400:
+                raise GCEVolumeException(
+                    "Cannot destroy volume {}. It is attached to an "
+                    "instance.".format(blockdevice_id)
```

Comment:
:question: Is this the only known reason for a 400 response?
Unless we're sure about that, perhaps we should include the original HTTP error in the raised exception.
Or log the exception.